### PR TITLE
sql: Remove redundant runSinglePrivilegeCheck grant check

### PR DESCRIPTION
**Pull request by ecwall**
refs https://github.com/cockroachdb/cockroach/issues/73129
    
HasPrivilege checks for privilege and grant permissions so runSinglePrivilegeCheck is no longer needed.

### CODE REVIEW

**Review by postamar — COMMENTED** [review]
> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75591)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75591#-MuOLNn6C-hIQn5V435t:-MuOLNn6C-hIQn5V435u:boymue6) ([raw file](https://github.com/cockroachdb/cockroach/blob/082134466946266fa1ef20c5230c43802275ccac/pkg/sql/sem/builtins/pg_builtins.go#L497)):*
> ```Go
> 		switch d {
> 		case tree.DBoolFalse:
> 		case tree.DBoolTrue, tree.DNull:
> ```

The handling of `tree.DNull` changed. Is this expected? Previously it was grouped with `tree.DBoolFalse` and would therefore not exit the function early.

Can `tree.DNull` arise at all in practice? If not, shouldn't it be handled by the `default` case?

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75591)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 497 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75591#-MuOLNn6C-hIQn5V435t:-MuQv6rX9BkDLq52NgXu:bno34z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/082134466946266fa1ef20c5230c43802275ccac/pkg/sql/sem/builtins/pg_builtins.go#L497)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

The handling of `tree.DNull` changed. Is this expected? Previously it was grouped with `tree.DBoolFalse` and would therefore not exit the function early.

Can `tree.DNull` arise at all in practice? If not, shouldn't it be handled by the `default` case?

</blockquote></details>

This logic is kind of hard to follow because of all the nested function calls and should be further simplified, but that requires resolving a circular dependency caused by adding `catalog` to the `EvalPlanner` API.

`check(priv)` can return the following
* `tree.DBoolFalse` - no privilege
* `tree.DBoolTrue` - privilege
* `tree.DNull` - the database object __specified by OID__ you are checking the privileges of does not exist
* `error w/ not found error code` - the database object __specified by name__ you are checking the privileges of does not exist

`tree.DBoolFalse` and `tree.DBoolTrue` depend on the privilege you are checking so they can be different for each `check` call - `tree.DBoolTrue` can be short circuited because `has_*_privilege` returns true if it has any specified privileges
`tree.DNull` and `error w/ not found error code` depend on the database object so the result will be the same for each `check` call which is why it can be short circuited

Resolving the circular dependency would let us split up checking if the object exists (only needs to happen once) with checking the privileges on it (happens once per specified privilege).

**Review by rafiss — APPROVED** [review]
lgtm!

> Reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75591#-:-MuS2ZCqBMBXQvCatDkk:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @postamar)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75591#-MuS2eZc9ZXzeXERpFmb):*
should this link to https://github.com/cockroachdb/cockroach/issues/66173 instead?

also, maybe it could say "fixes" instead of "refs"

**Review by ecwall — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75591)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @rafiss)

___
*[-- commits, line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75591#-MuS2eZc9ZXzeXERpFmb):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

should this link to https://github.com/cockroachdb/cockroach/issues/66173 instead?

also, maybe it could say "fixes" instead of "refs"
</blockquote></details>

Ah it should actually be this since thats where the redundant check came from originally: https://github.com/cockroachdb/cockroach/issues/73129

66173 contained the work required to be able to remove the check.

**Comment by ecwall** [comment]
bors r=rafiss

**Comment by ecwall** [comment]
bors r=rafiss
